runtime/domain.c

```diff
@@ -1486,25 +1494,77 @@ void caml_poll_gc_work(void)
 {
   CAMLalloc_point_here;
 
-  if (((uintnat)Caml_state->young_ptr - Bhsize_wosize(Max_young_wosize) <
-       (uintnat)Caml_state->young_start) ||
-      Caml_state->requested_minor_gc) {
+  caml_domain_state* d = Caml_state;
+
+  if ((uintnat)d->young_ptr - Bhsize_wosize(Max_young_wosize) <
+      (uintnat)d->young_trigger) {
+
+    if (d->young_trigger == d->young_start) {
+      /* Trigger minor GC */
+      d->requested_minor_gc = 1;
+    } else {
+      CAMLassert (d->young_trigger ==
+                  d->young_start + (d->young_end - d->young_start) / 2);
+      /* We have used half of our minor heap arena. Request a major slice on
+         this domain. */
+      d->requested_major_slice = 1;
+
+      uintnat old_value =
+        atomic_exchange (&caml_major_slice_epoch,
+        atomic_load (&caml_stat_minor_collections));
+      if (old_value != atomic_load (&caml_stat_minor_collections)) {
+        /* This domain is the first one to use up half of its minor heap arena
+           in this minor cycle. Trigger major slice on other domains. */
+        if (caml_plat_try_lock(&all_domains_lock)) {
+          /* Note that this interrupt is best-effort. If we get the lock,
+             then interrupt all the domains. If not, either some other domain
+             is calling for a stop-the-world section interrupting all the
+             domains, or a domain is being created or terminated. All of these
+             actions also try to lock [all_domains_lock] mutex, and the above
+             lock acquisition may fail.
+
+             If we don't get the lock, we don't interrupt other domains. This
+             is acceptable since it does not affect safety but only liveness --
+             the speed of the major gc. The other domains may themselves fill
+             half of their minor heap triggering a major slice, or do it right
+             after their next minor GC when they observe that their
+             domain-local [Caml_state->major_slice_epoch] is less than the
+             global one [caml_major_slice_epoch]. */
```

Comment:
IIUC there is some involved logic to make sure that other domains work enough for the major GC before being forced into a minor collection.

Here is another possible approach (this is for the record, as I had the chance to propose this to KC verbally already).
- Every time a minor collection is needed while the minor heap is half-empty, do a stop&copy collection into the free half of the minor heap. This reduces the amount of false promotions and (what is important here) avoids increasing the marking debt for the major GC.
- Every time a minor collection is needed before the major slice is done, simply transfer the marking debt to the next minor cycle.

Note: the major slice happens when the minor-heap is half-full, so the two trigger conditions coincide. Therefore, the major slice can be delayed, but the marking debt never accumulates.

In this approach, any additional work needed to keep up with phase changes (see ocaml/ocaml#11589) should be accounted for separately from the slice-at-middle rendez-vous.

Reply:
Hi @gadmm, I'm trying to understand how the proposal to perform stop & copy collection in the minor heap is solving the current problem. I'd think that is an orthogonal improvement. Perhaps your intention is to document future GC improvements. Not sure what the best place for this would be. Perhaps ocaml/RFCs? 

There are some (not insurmountable) challenges with implementing this copying collection for minor heap given that we may have pointers between the minor heap arenas of different domains. A parallel copying collector will have to synchronize to copy the objects to the `to_space`. While I've implemented simple Cheney copying collectors before, I don't know the state of the art in parallel copying collections. 

Coming back to the lines of code under discussion, in practice, I would think that the best-effort interruption above would work well. If the lock is not obtained, the worst that can happen is that the major slice is pushed to after the next minor collection on the other domains.

Reply:
Your current strategy is to trigger slices on all domains as soon as one domain reaches half of the minor heap. This is a bit heuristic, for instance due to scheduling hazards mutators on other domains could very well barely be restarting after the minor collection, in which case the splitting will not be effective in reducing latency. It might very-well work in practice, but I was pointing out that there is no need for a heuristic approach, since with the the semispace copying approach, the logic for when to do a slice is independent of the work of other domains.

I would not make this suggestion here if I thought it was as complex as you seem to think. I will be happy to discuss details elsewhere but ocaml/RFCs is probably not a good place for this (this concerns the implementation rather than a language change, and also some insiders discouraged in private its use so I do not know what to think about it).

I am not worried about the try_lock; have you considered accessing the interruptor via all_domains instead of stw_participants to avoid having to lock at all? (see this approach here: https://github.com/ocaml/ocaml/pull/11307/commits/dcb65204b9a098c713dcef54abf05752e7cbf6c1#diff-67115925103982a8ebeb085cfab5ef31a182c9a442bc51e053934364d3750dafR1507-R1517)

Reply:
> I would not make this suggestion here if I thought it was as complex as you seem to think.

As with anything in the GC implementation, my experience is that, for anything that is not trivial, there is typically a significant amount of implementation, debugging and tuning effort required. My approach has been to hack up a prototype and only test it on the best-case input for that design. If that is promising, then start implementing that feature for real. 

> I will be happy to discuss details elsewhere but ocaml/RFCs is probably not a good place for this (this concerns the implementation rather than a language change, and also some insiders discouraged in private its use so I do not know what to think about it).

I shall bring this up again with the caml-devel.

> have you considered accessing the interruptor via all_domains instead of stw_participants to avoid having to lock at all? (see this approach here: https://github.com/ocaml/ocaml/commit/dcb65204b9a098c713dcef54abf05752e7cbf6c1#diff-67115925103982a8ebeb085cfab5ef31a182c9a442bc51e053934364d3750dafR1507-R1517)

Is this safe? I am worried about races wrt other operations such as spawning and terminating domains. It felt uneasy to muck with their interruption flag and young limit. It would be nice to write down the guarantee for interruption as code comments along with a short account of why. What do you think? 

Given that triggering slices on other domains happens once every minor GC cycle, the current approach of taking the lock for interruption should not be expensive.

Reply:
The PR I pointed to (#11307) has to make a minor change to synchronize the read of `interruptor_word` with domain initialization. Modulo this change, this is safe because `interruptor_word` is set once and for all. I think #11307 documents enough the needed synchronisation but let me know if it is unclear (let's discuss it over there). Once #11307 is accepted, the function `caml_interrupt_all_for_signal` can be freely re-used for this purpose.

I don't have objections to the current heuristic for triggering slices nor to taking the lock, this is taking a good shot at a first implementation of the feature.